### PR TITLE
[custom_op] explicit autograd API

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #101824
* #101823
* #101822

This PR adds an explicit API for registering a backward formula for a
CustomOp. In the end state, we will likely have this explicit API and a
magic API (which is sugar on top of an explicit API), since different
parties of users prefer different ones.

Concretely, to define a backward formula for a CustomOp:
- a user must provide us a "save for backward" function that accepts
(inputs, output) and returns exactly what they want saved for backward
- a user must provide us a "backward" function that accepts
(ctx, saved, *grads) and returns us the grad_inputs. The grad_inputs
are returned as a dict mapping str to a gradient.
Please see the changes in custom_op_db.py for examples of the API.

There are a number of pieces to this PR and I'm happy to split it if it
helps. They are:
- The actual APIs for specifying the two functions
(impl_save_for_backward, impl_backward)
- The autograd kernel: we take the functions the user give us and
construct an autograd.Function object that we then register to
the Autograd dispatch key
- Indirection for the autograd kernel. We add a layer of indirection so
that one can swap out the autograd kernel. This is necessary because by
default, we register an "autograd not implemented" kernel as the
Autograd implementation but then swap it for the actual kernel when the
user provides it.

Test Plan:
- We apply this API to give backward formulas for things in
custom_op_db. We then hook up custom_op_db to the Autograd OpInfo tests.
- Various tests in test_python_dispatch.py to check error cases.

### CODE REVIEW

File: torch/_custom_op/autograd.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+        }
+    )
+    return generated_cls
+
+
+def namedtuple_args(schema, args):
+    assert isinstance(args, tuple)
+    attribs = [arg.name for arg in schema.arguments.flat_all]
+    name = str(schema.name) + "Args"
+    tuple_cls = namedtuple(name, attribs)
```
Comment (ezyang) [inline torch/_custom_op/autograd.py:120]: Do you want to cache this or something, so that repeated calls with the same schema return consistent named tuple objects?

**Review by ezyang — APPROVED** [review]
LGTM. But perhaps you want to have one of the autograd experts take a look over.

**Review by ezyang — APPROVED** [review]
LGTM. But perhaps you want to have one of the autograd experts take a look over.

File: torch/_custom_op/autograd.py (hunk 1 above)
Comment (albanD) [reply torch/_custom_op/autograd.py:120]: moving this to a helper with functools.cache() sounds good.

File: torch/_custom_op/autograd.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+            f"In the backward function defined for {forward_op} at "
+            f"{backward.location} using the CustomOp API, {what}")
+
+    if not isinstance(grad_inputs_dict, dict):
+        error(f"expected the output of the backward function to be a dict but "
+              f"got {type(grad_inputs_dict)}")
+
+    expected_keys = {arg.name for arg in forward_op._schema.arguments.flat_all
+                     if arg.type.is_tensor_like()}
+    actual_keys = {k for k, _ in grad_inputs_dict.items()}
```
Comment (albanD) [inline torch/_custom_op/autograd.py:137]: grad_inputs_dict.keys() ?
You can compare the result of this with the dict above or a set.

File: torch/_custom_op/impl.py [hunk 3]
```diff
@@ -417,6 +415,87 @@ def error_on_ctx():
...
+        In order for the CustomOp to work with autograd, you need to register
+        a backward formula. There are two pieces to this:
+        1. You must give us a function to specify what to save for backward.
+           Call this the "save for backward" function.
+        2. You must give us a function that computes gradients. Call this the
+           "backward" function.
+
+        Use `impl_save_for_backward` to define a "save for backward" function
+        that specifies what gets saved for backward. The function should accept
+        two arguments ``(inputs, output)`` and return the quantities to be saved
```
Comment (albanD) [inline torch/_custom_op/impl.py:454]: inputs, outputs ?

File: torch/_custom_op/impl.py [hunk 4]
```diff
@@ -417,6 +415,87 @@ def error_on_ctx():
...
+        - ``ctx`` is a context object where we may provide information
+        - ``saved`` is exactly what gets returned from the "save for backward"
+          function
+        - ``grads`` is one or more gradients. The number of gradients matches
+          the number of outputs of the CustomOp.
+
+        The backward function must return a dict that maps the name of
+        an input to the CustomOp to its corresponding gradient. All inputs that
+        were declared to be Tensors in the CustomOp definition must be accounted
+        for in the dict. The gradient may be a Tensor or None.
```
Comment (albanD) [inline torch/_custom_op/impl.py:471]: I'm not sure what's the benefit of using a dict here vs a tuple where there should be a 1-1 match anymore?
Is it just so that we don't need to pad with Nones anymore?

In the original design we were using dicts here for the registration API, it was a nice way to specify what each function was computing. But if this is doing eager evaluation is it still needed?

File: torch/testing/_internal/custom_op_db.py [hunk 5]
```diff
@@ -56,6 +66,22 @@ def numpy_mul_abstract(x, y):
...
+    saved['x_requires_grad'] = inputs.x.requires_grad
+    saved['y_requires_grad'] = inputs.y.requires_grad
+    # Optimization: only save what is necessary
+    saved['y'] = inputs.y if inputs.x.requires_grad else None
+    saved['x'] = inputs.x if inputs.y.requires_grad else None
+    return saved
+
+@numpy_mul.impl_backward()
+def numpy_mul_backward(ctx, saved, grad_out):
+    grad_x = grad_out * saved['y'] if saved['x_requires_grad'] else None
```
Comment (albanD) [inline torch/testing/_internal/custom_op_db.py:81]: The ctx should be telling you what to compute gradients for really.
This is fine as a test but we should not recommend users do this. We should just expose `needs_input_grad` on the ctx.

File: torch/_custom_op/autograd.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+                del guard
+
+            # We use the info about args to give better error messages in backward
+            args_info = namedtuple_args(
+                schema, pytree.tree_map(lambda arg: type(arg), args))
+
+            save_for_backward_fn_inputs = namedtuple_args(schema, args)
+            to_save = save_for_backward_fn(save_for_backward_fn_inputs, output)
+
+            save_pytree_for_backward(ctx, (to_save, args_info))
```
Comment (albanD) [inline torch/_custom_op/autograd.py:69]: nit we should be able to skip this when grad mode is disabled?

File: torch/_custom_op/autograd.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
+import torch
+import torch.utils._pytree as pytree
+from collections import namedtuple
+
+
+# NOTE [CustomOp autograd kernel indirection]
```
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/_custom_op/autograd.py:6]: This was meant to be the note, I didn't fill it out. I'll fill it out

File: torch/_custom_op/autograd.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+def autograd_kernel_indirection(custom_op):
+    # TODO(#101191): Use the actual C++ autograd not implemented fallback,
+    # or change the default autograd fallback to the autograd not implemented fallback.
+    def autograd_not_implemented(*args, **kwargs) -> None:
+        if pytree.tree_any(
+            lambda x: isinstance(x, torch.Tensor) and x.requires_grad, (args, kwargs)
```
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/_custom_op/autograd.py:12]: Yes, good point

File: torch/_custom_op/autograd.py [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+        try:
+            return custom_op(*args, **kwargs)
+        finally:
+            del guard
+
+    def inner(*args, **kwargs):
+        if custom_op._has_impl('autograd'):
+            kernel = custom_op._get_impl('autograd').func
+            return kernel(*args, **kwargs)
+        if custom_op._has_impl('save_for_backward') or custom_op._has_impl('backward'):
```
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/_custom_op/autograd.py:25]: That is the case, yes. I'll add a comment

File: torch/_custom_op/autograd.py (hunk 2 above)
Comment (zou3519) [reply torch/_custom_op/autograd.py:137]: I was a bit scared that the type of `dict.keys()` is not actually a dict or a set, but I'll try it out

File: torch/_custom_op/impl.py (hunk 3 above)
Comment (zou3519) [reply torch/_custom_op/impl.py:454]: No, this is actually `inputs, output`:
- If the output is a single Tensor, then `output` is a single Tensor
- If the output is a Tuple of Tensors, then `output` is a Tuple of Tensors
- If the output is a tuple of a single Tensor, then `output` is a Tuple of a single Tensor
I'm open to changing it (though, this is how setup_context works as well).

File: torch/_custom_op/impl.py (hunk 4 above)
Comment (zou3519) [reply torch/_custom_op/impl.py:471]: > I'm not sure what's the benefit of using a dict here vs a tuple where there should be a 1-1 match anymore?
Is it just so that we don't need to pad with Nones anymore?

The main motivation is so that we don't need to pad with Nones anymore. Padding with Nones was a bit annoying to do, especially with autograd.Function that had a long list of inputs. I can be convinced to go back to the 1:1 match.

> In the original design we were using dicts here for the registration API, it was a nice way to specify what each function was computing. But if this is doing eager evaluation is it still needed?

We can still have a separate derivatives.yaml-like API for this, since there are separate technical benefits to the separate derivatives.yaml declarations vs the autograd.Function style. That one can look something like:

```py
myop.def_autograd({
   'x': ...
   'y': ...
})

```

It may be confusing to use dicts in both places, though.

File: torch/testing/_internal/custom_op_db.py (hunk 5 above)
Comment (zou3519) [reply torch/testing/_internal/custom_op_db.py:81]: Agree, I didn't add needs_input_grad in this PR because it was getting long.

File: torch/_custom_op/autograd.py (hunk 6 above)
Comment (zou3519) [reply torch/_custom_op/autograd.py:69]: Sure

File: torch/testing/_internal/custom_op_db.py (hunk 5 above)
Comment (zou3519) [reply torch/testing/_internal/custom_op_db.py:81]: I can add it as a follow-up, unless you want to see it in this PR?

File: torch/_custom_op/impl.py (hunk 3 above)
Comment (albanD) [reply torch/_custom_op/impl.py:454]: I'm curious:
- Why do you prefer this semantic?
- What is the semantic used in our other similar functions? (hooks always wrap in a tuple IIRC?)

File: torch/_custom_op/autograd.py [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+# NOTE [CustomOp autograd kernel indirection]
+def autograd_kernel_indirection(custom_op):
+    # TODO(#101191): Use the actual C++ autograd not implemented fallback,
+    # or change the default autograd fallback to the autograd not implemented fallback.
+    def autograd_not_implemented(*args, **kwargs) -> None:
+        if pytree.tree_any(
+            lambda x: isinstance(x, torch.Tensor) and x.requires_grad, (args, kwargs)
+        ):
+            raise RuntimeError("Autograd has not been implemented for operator")
+        guard = torch._C._AutoDispatchBelowAutograd()
```
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/_custom_op/autograd.py:15]: https://github.com/pytorch/pytorch/issues/101191

File: torch/_custom_op/impl.py (hunk 4 above)
Comment (albanD) [reply torch/_custom_op/impl.py:471]: > We can still have a separate derivatives.yaml-like API for this

I was not suggesting that no. I agree we shouldn't mix them.

> The main motivation is so that we don't need to pad with Nones anymore.

Sounds fair!
We can always add tuple support if some people prefer later on.

File: torch/testing/_internal/custom_op_db.py (hunk 5 above)
Comment (albanD) [reply torch/testing/_internal/custom_op_db.py:81]: No need to add it here as long as this is not in any user-facing example!

File: torch/_custom_op/impl.py (hunk 3 above)
Comment (zou3519) [reply torch/_custom_op/impl.py:454]: > Why do you prefer this semantic?

- It's easier to explain that the `output` is exactly the output of the `forward()`, rather than "we wrap the output in a Tuple if it is not a Tuple".
- It avoids ambiguity: If we support returning TensorList (which we likely will in the future, I have not enabled it in this PR), and we have `foo(Tensor x) -> TensorList`, then there is the question of if `outputs` is a Tuple[List[Tensor]] or just a List[Tensor].

> What is the semantic used in our other similar functions? (hooks always wrap in a tuple IIRC?)

- nn.Module `register_forward_hook` passes the [output as-is without wrapping](https://gist.github.com/zou3519/822505b737f8bc0646e45680bc7798cc)
- `register_backward_hook` (which requires hook of signature `(mod, grad_input, grad_output)` wraps grad_output as a Tuple.
- autograd.Function's forward distinguishes between `return (x,)` and `return x`
- autograd.Function's backward does not distinguish between `return grad` and `return (grad,)`

Based on those examples it seems like for the forward pass we are generally careful that the types match, but when it comes to gradients we generally wrap things in a Tuple to make things simpler.

Comment (albanD) [reply torch/_custom_op/impl.py:454]: Ok! Keeping it as is sounds good then!

File: torch/_custom_op/autograd.py [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+                        ctx.mark_non_differentiable(output)
+
+            return output
+
+        def backward(ctx, *grads):
+            saved, args_info = unpack_saved(ctx)
+            # There is nothing on the ctx object for now, it is just there so
+            # that we can add additional things in the future.
+            inner_ctx = object()
+            grad_inputs_dict = backward_fn(inner_ctx, saved, *grads)
```
Comment (soulitzer) [inline torch/_custom_op/autograd.py:90]: Curious what the API would look like with the implicit autograd API, are there any bc-breaking things we should be worried about when we transition to that one?

File: torch/_custom_op/autograd.py [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+
+            save_for_backward_fn_inputs = namedtuple_args(schema, args)
+            to_save = save_for_backward_fn(save_for_backward_fn_inputs, output)
+
+            save_pytree_for_backward(ctx, (to_save, args_info))
+
+            # Output must be one or more Tensors, no TensorList (yet)
+            if output_differentiability is not None:
+                if isinstance(output, tuple):
+                    assert len(output_differentiability) == len(output)
```
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/_custom_op/autograd.py:74]: Yeah, this is checked in `impl_backward` (and we have tests)

**Review by zou3519 — COMMENTED** [review]
.

File: torch/_custom_op/autograd.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/_custom_op/autograd.py:25]: > Is it so that we can avoid passing save_for_backward through the decorator?

Yes. It felt strange that we were passing a function through a decorator, so I split it into two decorators.

> The argument for combining it is that since we always need both, might as well have them registered in the same place.

Yeah I agree that this is an argument for the opposite.

File: torch/_custom_op/autograd.py (hunk 11 above)
Comment (zou3519) [reply torch/_custom_op/autograd.py:90]: By "implicit" do you mean "magic"?

We can just name the magic API something different; then there are no BC-breaking changes. But I might be misunderstanding the question
```py
@myop.impl_backward_unified(output_differentiability=...)
def backward(grad, x):
    return grad * x.cos()
```

File: torch/_custom_op/autograd.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/_custom_op/autograd.py:12]: It is a good idea to give people an easy way to do this (so they don't need to define their save_for_backward function and their backward function in order to say that their function is not differentiable). I can address that as a follow up, unless you want to see it in this PR?

I'm not really sure what this should look like yet: we could ask the author to give us a `differentiable` flag at creation of the custom_op, or we could ask them to `impl_backward(None)`, or something else.

File: torch/_custom_op/autograd.py (hunk 2 above)
Comment (zou3519) [reply torch/_custom_op/autograd.py:137]: Yes, this works

File: torch/_custom_op/autograd.py (hunk 6 above)
Comment (zou3519) [reply torch/_custom_op/autograd.py:69]: It is a little complicated to do this because autograd.Function's forward sets grad_mode to False. This means that we would need to shepherd the knowledge of if grad mode is disabled or not from somewhere else.

Since this is a nit I'm just going to add this to the wishlist.

File: torch/_custom_op/autograd.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/_custom_op/autograd.py:12]: Follow up PR sounds ok!

> I'm not really sure what this should look like yet: we could ask the author to give us a differentiable flag at creation of the custom_op, or we could ask them to impl_backward(None), or something else.

My vote would be for having a `differentiable` flag at custom_op creation since whether the op is differentiable or not seems like a property of the op itself.

File: test/test_python_dispatch.py [hunk 13]
```diff
@@ -785,6 +785,271 @@ def foo_impl(x):
...
+        @custom_op(f'{TestCustomOp.test_ns}::foo')
+        def foo(xs: Sequence[torch.Tensor]) -> torch.Tensor:
+            ...
+
+        with self.assertRaisesRegex(RuntimeError, "output_differentiability"):
+            @foo.impl_backward(output_differentiability=True)
+            def foo_backward(ctx, saved, grad):
+                return {'xs': None}
+
+    def test_backward_output_differentiability_elements(self):
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply test/test_python_dispatch.py:1043]: Renaming also sounds good!

File: torch/_custom_op/autograd.py (hunk 11 above)
Comment (soulitzer) [reply torch/_custom_op/autograd.py:90]: Oh yeah thats what I meant. That plan sounds good.

File: test/test_python_dispatch.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (zou3519) [reply test/test_python_dispatch.py]: Renamed

File: torch/_custom_op/impl.py [hunk 14]
```diff
@@ -417,6 +415,90 @@ def error_on_ctx():
...
+
+        TODO(rzou): Add example when this PR is closer to landing.
```
Comment (zou3519) [inline torch/_custom_op/impl.py:476]: Sorry, I remembered this after I typed the merge command. Will do as a follow-up.

**Comment by vadimkantorov** [comment]
Regarding the syntax, could `custom_op` be combined somehow with forward impl?
currently:
```python
        @custom_op(f'{TestCustomOp.test_ns}::foo')
        def foo(x: torch.Tensor) -> torch.Tensor:
            ...

        @foo.impl(['cpu', 'cuda'])
        def foo_impl(x):
            return x.sin()

        @foo.impl_backward()
        def foo_backward(ctx, saved, grad):
            return grad * saved.cos()
```

proposed:
```python
        @custom_op_with_forward(f'{TestCustomOp.test_ns}::foo', device_types = ['cpu', 'cuda'])
        def foo(x : torch.Tensor) -> torch.Tensor:
            return x.sin()

        @foo.impl_backward()
        def foo_backward(ctx, saved, grad):
            return grad * saved.cos()
```
